File: CPP_Solutions/Longest_Common_Subsequence/Print_SCS.cpp
#include<iostream>
#include<algorithm>
#include<cstring>
using namespace std; 

string print_SCS(char x[], char y[], int n, int m){
    
    // apply LCS
    int t[n+1][m+1];
    
    for(int i = 0; i < n + 1; i++)
        t[i][0] = 0;
    for(int j = 0; j < m + 1; j++)
        t[0][j] = 0;
        
    for(int i = 1; i < n + 1; i++)
        for(int j = 1; j < m + 1; j++)
            t[i][j] = (x[i-1] == y[j-1]) ? (1 + t[i-1][j-1]) : std::max(t[i-1][j], t[i][j-1]);
    // value for t[n][m] has been obtained by applying LCS
    string scs;
    int i = n , j = m ;
    while(i > 0 && j > 0){
        if(x[i-1] == y[j-1]){
            scs.push_back(x[i-1]);
            i--;
            j--;
        }
        else if(t[i-1][j] > t[i][j-1]){
            scs.push_back(x[i-1]);
            i--;
        }
        else {
            scs.push_back(y[j-1]);
            j--;
        }
    }
    while(i > 0){ // j  has already become zero
        scs.push_back(x[i-1]);
        i--;
    }
    while(j > 0){// i has already become zero
        scs.push_back(y[j-1]);
        j--;
    }
    reverse(scs.begin(),scs.end());
    return scs;
}

int main() 
{
    char x[] = "acbcf";
    char y[] = "abcdaf";
    int n = (sizeof(x) / sizeof(x[0])) - 1;
    int m = (sizeof(y) / sizeof(y[0])) - 1;
    cout << print_SCS(x,y,n,m);
    return 0;
}

File: Python_Solutions/Longest_Common_Subsequence/Print_SCS.py
def print_SCS(x,y,n,m):
    t = [[0 for j in range(m+1)] for i in range(n+1)] # Initialise matrix with zero

    for i in range(1, n+1):
        for j in range(1, m+1):
            t[i][j] = (1 + t[i-1][j-1]) if(x[i-1] == y[j-1]) else (max(t[i-1][j], t[i][j-1]))

    # Value for t[n][m] obtained

    scs = ""
    i = n
    j = m
    while i > 0 and j > 0:
        if(x[i-1] == y[j-1]):
            scs += x[i-1]
            i-=1
            j-=1
        elif(t[i-1][j]>t[i][j-1]):
            scs += x[i-1]
            i-=1
        else:
            scs += y[j-1]
            j-=1
    
    while i > 0: # j has already become zero
        scs += x[i-1]
        i-=1
    while j > 0: # i has already become zero
        scs += y[j-1]
        j-=1
    scs = list(scs)
    scs.reverse()
    return "".join(scs) # since scs stores the each character as elements of an array/list

x = "acbcf"
y = "abcdaf"
print(print_SCS(x,y,len(x),len(y)))